src/usr/ipmi/ipmifruinv.C

```diff
@@ -522,6 +667,201 @@ errlHndl_t isdimmIpmiFruInv::addVpdData(std::vector<uint8_t> &io_data,
+            // density = 256 Mb* (2 ^ density) / 8 = 32 MB * (2 ^ density)
+            const uint32_t l_realDensity = (JEDEC_DENSITY_MIN_VALUE / 8) << l_density;
+
+            // Calculate the Primary Bus Width
+            // JEDEC Standard No. 21-C. Page 4.1.2.12 – 15
+            // b00 - 8 bits ... b11 - 64 bits. All others reserved.
+            const uint32_t l_realBusWidth = JEDEC_MEMORY_BUS_WIDTH_PRI_MIN_VALUE << l_busWidthPri;
+
+            // Calculate the SDRAM Device Width
+            // JEDEC Standard No. 21-C. Page 4.1.2.12 – 14
+            // b00 - 4 bits ... b11 - 32 bits. All others reserved.
+            const uint32_t l_realDevWidth = JEDEC_DRAM_WIDTH_MIN_VALUE << l_dramWidth;
+
+            // Calculate the Number of Package Ranks per DIMM
+            // JEDEC Standard No. 21-C. Page 4.1.2.12 – 14
+            // b00 - 1 package rank ... b11 - 4 package ranks. All others reserved.
+            const uint32_t l_realRanks = JEDEC_RANKS_MIN_VALUE + l_ranks;
+
+            // Calculate the Module Capacity (in GiB >> 10) according to the formula
+            // from the JEDEC Standard specification No. 21-C. Page 4.1.2.12 – 15
+            l_capacityGiB = (l_realDensity * (l_realBusWidth / l_realDevWidth) * l_realRanks) >> 10;
+        }
+
+        // Construct detailed string description
+        char l_desc[IPMIFRUINV::MAX_ASCII_FIELD_SIZE] = { 0 };
+        // Always store last null-termination symbol
+        const size_t l_descSz = sizeof(l_desc) - 1;
+
+        const char* l_decoded;
+
+        l_decoded = jedecDecode(SPD::BASIC_MEMORY_TYPE, l_basicType);
+        if (l_decoded)
+        {
+            strncpy(l_desc, l_decoded, l_descSz);
+        }
+        else
+        {
+            snprintf(l_desc, l_descSz, "N/A (%02x)", l_basicType);
+        }
+
+        l_decoded = jedecDecode(SPD::TCK_MIN, l_tckMin);
+        if (l_decoded)
+        {
+            strncat(l_desc, "-", l_descSz);
+            strncat(l_desc, l_decoded, l_descSz);
+        }
+
+        if (l_capacityGiB)
+        {
+            char l_buff[16] = { 0 };
+            snprintf(l_buff, sizeof(l_buff) - 1, " %iGiB", l_capacityGiB);
+            strncat(l_desc, l_buff, l_descSz);
+        }
+
+        l_decoded = jedecDecode(SPD::MODULE_MEMORY_BUS_WIDTH_PRI, l_busWidthPri);
+        if (l_decoded)
+        {
+            strncat(l_desc, " ", l_descSz);
+            strncat(l_desc, l_decoded, l_descSz);
+        }
+
+        l_decoded = jedecDecode(SPD::MODULE_MEMORY_BUS_WIDTH_EXT, l_busWidthExt);
+        if (l_decoded)
+        {
+            strncat(l_desc, " ", l_descSz);
+            strncat(l_desc, l_decoded, l_descSz);
+        }
+
+        l_decoded = jedecDecode(SPD::MODULE_TYPE, l_moduleType);
+        if (l_decoded)
+        {
+            strncat(l_desc, " ", l_descSz);
+            strncat(l_desc, l_decoded, l_descSz);
+        }
+
+        // Put detailed description into the message buffer
+        const size_t l_len = strlen(l_desc);
```

Comment:
According to IPMI FRU specification, value 0 for the length means that the fields is empty, while value 1 indicates 'end-of-fields'. Thus, value 1 is always reserved and can't be used. Which means that the minimum size of a string is 2. This is not checked here.

Also, the maximum length of plain ASCII string in IPMI is 31 bytes. This is 1 byte shorter than, for instance, "DDR4-2400 16GiB 64-bit ECC RDIMM". Things like "LPDDR4-2400 16GiB 64-bit ECC SORDIMM 72b" will get even more fiercely truncated. Something must be invented to overcome that. I'd propose converting the string to all-uppercase and encoding it using 6-bit ASCII (see FRU spec). That would give us 41 bytes max. We could also save on using non-SI units for size (`GB` instead of `GiB`), on saying `64B` instead of `64-BIT`, and on shortening long module types to, say these:

- `MINIRDIMM`
- `MINIUDIMM`
- `SORDIMM72`
- `SODIMM16`
- `SODIMM32`

That way the longest string will be `LPDDR4-2400 16GB 32B NON-ECC MINIRDIMM` which is 38 characters.

It must be checked though whether or not the BMC firmwares can properly decode 6-bit ASCII.

If we ditch Mini and SO modules, that are, AFAIK, rarely used in OpenPOWER systems, and also use empty string instead of 'NON-ECC', then we can fit into 31 bytes with the longest `LPDDR4-2400 16GB 32b ECC LRDIMM`, and that's the good old plain ASCII+Latin1 encoding.

Reply:
If we use something other than straight ascii, what are the chances that the BMC will actually display it in any useful way?  Might need to ask in the openbmc forum what they support.  Have you by chance tried this change on a system running OpenBMC (or SuperMicro or AMI)?  Where/how is the data displayed?

Reply:
OpenBMC always shows manufacture and pretty names as text.
We have checked this solution on our VESNIN server.

WebUI example:
![screenshot_2018-09-24 webui](https://user-images.githubusercontent.com/1169976/45941317-4709ad80-bfe6-11e8-85ac-143bfe3f4a43.png)

Inventory list:
![screenshot_2018-09-24 inventory](https://user-images.githubusercontent.com/1169976/45941320-4b35cb00-bfe6-11e8-8e46-dd48fc49e9dc.png)

Reply:
@artemsen, I believe @dcrowell77 was referring to my proposal to use IPMI FRU's 6-bit ASCII encoding. I have checked and am quite sure now that openbmc/ipmi-fru-parser does not support any encodings beyond binary (code 0) and plain ASCII (code 3 / 0xC0). Can't say for AMI. We of course could add support for 6-bit ASCII to openbmc, but hardly to MegaRAC or any other commercial products.

This means that our best way is to use plain ASCII and ditch support for the uncommon module types.

Reply:
> Also, the maximum length of plain ASCII string in IPMI is 31 bytes

Mea culpa. I misread the spec. The type/length byte has 6 bits for data length, which gives maximum string length of 63, not 31. I don't know where I took that '31' from. Nothing needs to be done in regard to string length.